### ISSUE TITLE
Support for continuous (autorun) tests

### ISSUE THREAD

**Issue by connor4312** [microsoft/vscode#134941]
A popular feature in test runners is re-running tests automatically as source code changes. I this should consider recognizing API, since the UI will have several specific behaviors for this:

- A profile-aware "toggle" button to start and stop running common across extensions and implementations
- Different behavior around when or if the peak automatically opens and state is followed in the test explorer when under an auto-run

While the former can be done through existing extension APIs such as menu contributions, it may be worth having official UI for this to provide a consistent experience, especially when multiple types of tests (and therefore extensions) are present.

I think this can be implemented in a fairly straightforward way with the addition of a `supportsContinuousRun` on the profile, and a boolean indicator on TestRuns indicating whether they should be continuous. I prefer the term "continuous" over "auto run" since saying "this is a continuous test run" brings in the temporal nature that "this is an autorun" does not.

<details>
<summary>API Proposal</summary>

```ts
	export interface TestRunProfile {
		/**
		 * Whether this profile supports continuous running of requests. If so,
		 * then {@link TestRunRequest.continuous} may be set to `true`. Defaults
		 * to false.
		 */
		supportsContinuousRun: boolean;

		/**
		 * Handler called to start a test run. When invoked, the function should call
		 * {@link TestController.createTestRun} at least once, and all test runs
		 * associated with the request should be created before the function returns
		 * or the returned promise is resolved.
		 *
		 * If {@link supportsContinuousRun} is set, then {@link TestRunRequest2.continuous}
		 * may be `true`. In this case, the profile should observe changes to
		 * source code and create new test runs by calling {@link TestController.createTestRun},
		 * until the cancellation is requested on the `token`.
		 *
		 * @param request Request information for the test run.
		 * @param cancellationToken Token that signals the used asked to abort the
		 * test run. If cancellation is requested on this token, all {@link TestRun}
		 * instances associated with the request will be
		 * automatically cancelled as well.
		 */
		runHandler: (request: TestRunRequest, token: CancellationToken) => Thenable<void> | void;

		// ...
	}

	export interface TestController {
		/**
		 * Creates a profile used for running tests. Extensions must create
		 * at least one profile in order for tests to be run.
		 * @param label A human-readable label for this profile.
		 * @param kind Configures what kind of execution this profile manages.
		 * @param runHandler Function called to start a test run.
		 * @param isDefault Whether this is the default action for its kind.
		 * @param tag Profile test tag.
		 * @param supportsContinuousRun Whether the profile supports continuous running.
		 * @returns An instance of a {@link TestRunProfile}, which is automatically
		 * associated with this controller.
		 */
		createRunProfile(label: string, kind: TestRunProfileKind, runHandler: (request: TestRunRequest, token: CancellationToken) => Thenable<void> | void, isDefault?: boolean, tag?: TestTag, supportsContinuousRun?: boolean): TestRunProfile;

		// ...
	}

	export class TestRunRequest {
		/**
		 * Whether the profile should run continuously as source code changes. Only
		 * relevant for profiles that set {@link TestRunProfile.supportsContinuousRun}.
		 */
		readonly continuous?: boolean;

		/**
		 * @param tests Array of specific tests to run, or undefined to run all tests
		 * @param exclude An array of tests to exclude from the run.
		 * @param profile The run profile used for this request.
		 * @param continuous Whether to run tests continuously as source changes.
		 */
		constructor(include?: readonly TestItem[], exclude?: readonly TestItem[], profile?: TestRunProfile, continuous?: boolean);

		// ...
	}
```

</details>

**Comment by JustinGrote** [comment]
For Pester Tests, currently I control this with a AutoRun setting, so it makes perfect sense to put it in profiles. 

One issue that has come up is, when autorunning, it is possible to save a file and commit with the test pane and test file closed, and have no idea what you changed just failed the test, there's no indicator in this situation

Thoughts:
1. If the ability to "alias" a test e.g. put a run button next to the function it references, and have that update, that would provide a good indicator, but still not useful if the user scrolls away after saving and doesn't notice. Maybe some sort of coverage API integration.
2.  Maybe allow a configurable setting to report test errors to the "Problems" pane.
3. Have a new editor tab indicator similar to the git "modified/error/etc." indicators that specifically notes a test failed here

**Comment by ashgti** [comment]
With this API, who is responsible for detecting test run trigger events? Is the implication that the extension should run on FS save or on editor text change?

Also, what is the flow of the `TestRunProfile.runRequest`, does it get called one time once continuous mode is enabled and then the cancellation token is cancelled when the mode is disabled or does it get called once per document save event?

**Comment by connor4312** [comment]
> Is the implication that the extension should run tests on FS save or on editor text change?

That's correct

> Also, what is the flow of the TestRunProfile.runRequest, does it get called one time when continuous mode is enabled and then the cancellation token is cancelled when the mode is disabled or does it get called once per document save event?

Once when enabled

**Comment by matepek** [comment]
+1

**Comment by JDaance** [comment]
I'm looking at implementing auto run right now and have a bit of trouble figuring out the best way. As it stands now, should I create a new run profile called continous or something, and simply never "return"? Then it can be cancelled by cancellation token?

**Comment by connor4312** [comment]
You still want to call [TestController.createTestRun](https://github.com/microsoft/vscode/blob/db79f4c92a9e34f3ff10d3a826cf4ebc295beb2c/src/vscode-dts/vscode.d.ts#L14639) whenever you start running tests. Note that you can just `new` your own `TestRunRequest`. Then use existing contribution points to add your own auto run toggle UI, for example a status bar contribution.

**Comment by JDaance** [comment]
> You still want to call [TestController.createTestRun](https://github.com/microsoft/vscode/blob/db79f4c92a9e34f3ff10d3a826cf4ebc295beb2c/src/vscode-dts/vscode.d.ts#L14639) whenever you start running tests -- on file save, for example. Note that you can just `new` your own `TestRunRequest`, it doesn't have to come from a `runHandler` invokation. Then use existing contribution points to add your own auto run toggle UI, for example a status bar contribution.

Ok gotcha, it would be nice to have the auto run toggle in the test explorer somewhere, but I guess thats what this issue is about. 

I am a bit confused about the settings available for the testing feature concerning auto run. I guess that these settings currently do not have any effect?

**Comment by connor4312** [comment]
>  I guess that these settings currently do not have any effect?

That's right

**Comment by JDaance** [comment]
![image](https://user-images.githubusercontent.com/1148112/153062891-b4f51a4b-bf74-4d00-ad87-90b0c327855e.png)
![image](https://user-images.githubusercontent.com/1148112/153062930-0763a8ce-4159-4801-b526-1ff331d90957.png)


Poor mans autorun toggle, until better support is available :)

Turns out I can change the label if I keep the reference

**Comment by dogweather** [comment]
> > I guess that these settings currently do not have any effect?

> That's right

Oy, why? I know I don't have a vote, but if I did, I'd vote against features that fail silently.

**Comment by B1773rm4n** [comment]
How often this will get delayed again?

**Comment by JustinGrote** [comment]
> I'm desperate for this!
You can implement it in a test adapter even if it doesn't exist in the core UI. My Pester Tests extension has an option to run every time you save for instance.

**Comment by ioquatix** [comment]
I've experimented with this and I found the way the UI pops up in the code a little clunky and hard to use while editing the text (if autosave is enabled or you trigger based on document edit/update).

I'd love to see us iterate on the way the UI provides feedback of test failure, both when editing the test, and when editing code that causes a test to fail.

**Comment by matepek** [comment]
I'm still rooting for this feature ^^

**Comment by karamfil** [comment]
Go Go Guys! You can do it and deliver it in November :)

**Comment by ioquatix** [comment]
I'm waiting on the sidelines for when I can roll out improved support in my extension! I'm excited!

**Comment by connor4312** [comment]
Proposed API for autorun will be available on Insiders early next week. PR with more info here: https://github.com/microsoft/vscode/pull/170770

Examples of adopting the change:

- Sample test provider: https://github.com/microsoft/vscode-extension-samples/commit/978f6538bd646f6fe140ac392386baaa00db348a
- Selfhost test provider: https://github.com/microsoft/vscode-selfhost-test-provider/pull/17/files

**Comment by ioquatix** [comment]
@connor4312 what kind of feedback are you looking for?

**Comment by connor4312** [comment]
You can give it a try with your extension and see if the overall API flow and scenario fits your needs. If not, you run into a bug, or find something that can be improved, let me know 🙂 

The UI changes I'm thinking of are mostly around reorganizing the action buttons, since the view title is getting a bit cluttered, especially with any extension-contributed actions that might appear. This was one rough draft of a reworked version that I put together and that I'll discuss with the team:


![image](https://user-images.githubusercontent.com/2230985/211176870-9db0a333-a885-4093-86e9-c841ecc18788.png)

**Comment by jabbera** [comment]
Having a count of failed/passed test and bottom with a background that changes to red when you have a failed test would be really nice.

**Comment by ioquatix** [comment]
For test statistics, I recommend:

> x passed, y failed, z errored, w skipped.

The text should be colored, green, red, red and yellow accordingly.

**Comment by matepek** [comment]
Hello @connor4312 ,

If I understand correctly one can just select a run profile to "autorun" but cannot select a set of tests.
We are hoping it will be supported in the future but I'm not sure how the current api supports that.
Maybe [request: vscode.TestRunRequest2](https://github.com/microsoft/vscode-extension-samples/commit/978f6538bd646f6fe140ac392386baaa00db348a#diff-2d9e354ff5c71e7317d87d1aefa9ea0bc7cb4ebb05b2841484af5a0bcddce120R9) can return with the `included` tests..
And then we just have to check that the modified file is relates to the `included` `TestItem`s or not.

What do you think?

**Comment by connor4312** [comment]
`vscode.TestRunRequest2` extends `vscode.TestRunRequest`, so also has "include" files. I just haven't added UI for that yet. Thanks for letting me know it's something you're interested in 🙂

**Comment by ioquatix** [comment]
I was under the impression that it might be possible to provide per-test coverage, so that we know what tests to run when a specific line of code is changed.

**Comment by connor4312** [comment]
> I was under the impression that it might be possible to provide per-test coverage, so that we know what tests to run when a specific line of code is changed.

This is up to extensions / test runners to provide. VS Code itself is not aware of your dependency tree or runtime environment. If the test runner you're working with generates per test coverage, that could be a reasonable way to figure out how to map code to tests.

**Comment by matepek** [comment]
I think I found an issue: debugging my extension with the insider the "Test Output" is only:
```
The test run did not record any output.
```
instead of the one I generated.

**Comment by ioquatix** [comment]
I've been playing around with triggering a test run when files are modified.

This UI pops up while I'm editing the test, which can be clunky:

<img width="960" alt="image" src="https://user-images.githubusercontent.com/30030/219998271-38c3ea41-b39f-4ca8-9a23-0849561fa882.png">

It would be good to think about a way to make sure the UI and test feedback is compatible with real-time editing of the test.

**Comment by connor4312** [comment]
Thanks, will look at that.

I have been out of office this month, so I will look to finalize this API (and clean up polish items as mentioned) in March

**Comment by ioquatix** [comment]
If you'd like me to walk through how my particular integration is going, and we can check the weak points and strong points, let me know :)
